### ISSUE TITLE
feat(distill): add hidden-state, quantization-aware, and online distillation strategies

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#451]
## Summary

Add 3 missing distillation strategies to achieve parity with DistillKit.

### Hidden State Distillation (P1)

Match intermediate hidden state representations between teacher and student
using learned linear projections to bridge dimensionality differences.

```bash
apr distill teacher.apr --student student.apr \
  --strategy hidden-state \
  --hidden-layer-map "0:0,4:1,8:2,12:3" \
  --data train.jsonl -o distilled.apr
```

**Parameters**:
- `--hidden-layer-map`: Teacher:Student layer index pairs
- `--hidden-weight`: Weight of hidden state loss vs logit loss (default: 1.0)
- `--projection-init`: Initialization for projection layers (default: xavier)

**DistillKit finding**: Logit-based consistently outperforms hidden-state
across benchmarks, but hidden-state can be superior for architecturally
diverse teacher-student pairs.

### Quantization-Aware Distillation (P2)

Student trained with simulated quantization noise so the final quantized
model retains more quality than post-training quantization alone.

```bash
apr distill teacher.apr --student student.apr \
  --strategy quant-aware \
  --target-quant q4k \
  --data train.jsonl -o distilled-q4k.apr
```

**Parameters**:
- `--target-quant`: Target quantization scheme (int4, int8, q4k, q6k)
- `--quant-noise-scale`: Simulated quantization noise amplitude

### Online Distillation (P2)

Concurrent teacher inference during student training — no precompute step.
Teacher and student run simultaneously on separate GPUs.

```bash
apr distill teacher.apr --student student.apr \
  --strategy online \
  --teacher-gpu 0 --student-gpu 1 \
  --data train.jsonl -o distilled.apr
```

**Trade-offs**: Requires 2x GPU memory but eliminates disk space for
cached logits and simplifies the pipeline to a single command.

## Specification

See `docs/specifications/components/distill.md` §5.

## Testing

- Hidden state: verify projection layer dimensions match
- Quant-aware: verify quantized student quality vs post-training quantization
- Online: verify parity with two-stage offline distillation

### Contract: `distillation-strategies-v2.yaml`

Extends knowledge distillation beyond logit-based (already covered by cross-entropy contracts).

```yaml
metadata:
  description: "Extended distillation — hidden state, quantization-aware, online"
  references:
    - "Romero et al. (2015) FitNets: Hints for Thin Deep Nets"
    - "Arcee AI (2024) DistillKit"
  depends_on:
    - "cross-entropy-kernel-v1"
    - "softmax-kernel-v1"

equations:
  hidden_state_loss:
    formula: "L_hidden = Σ ||proj(h_s^l) - h_t^{f(l)}||²"
    domain: "h_s = student hidden, h_t = teacher hidden, f = layer mapping"
    invariants:
      - "L_hidden >= 0 (MSE is non-negative)"
      - "proj dimensions: student_dim → teacher_dim"
      - "proj is learned (not fixed)"

  quant_aware_loss:
    formula: "L_qa = KL(softmax(t_logits/T), softmax(Q(s_logits)/T)) + noise_reg"
    invariants:
      - "Q() simulates target quantization during forward pass"
      - "Noise regularization prevents overfitting to quantization grid"
      - "L_qa >= 0"

  online_parity:
    formula: "L_online(batch) == L_offline(batch) for same teacher"
    invariants:
      - "Online and offline produce identical logits for same input"
      - "Teacher weights frozen in both modes"

proof_obligations:
  - type: bound
    property: "Hidden state loss non-negativity"
    formal: "L_hidden >= 0 (MSE)"
  - type: invariant
    property: "Projection dimension correctness"
    formal: "proj.weight.shape == (teacher_dim, student_dim)"
  - type: equivalence
    property: "Online/offline parity"
    formal: "teacher_logits_online == teacher_logits_offline for same input"
  - type: invariant
    property: "Teacher frozen in all modes"
    formal: "teacher weights unchanged after distillation"
  - type: bound
    property: "Quant-aware loss non-negativity"
    formal: "L_qa >= 0"

falsification_tests:
  - id: FALSIFY-DIST-001
    rule: "Hidden loss non-negative"
    prediction: "L_hidden >= 0 for random hidden states"
    if_fails: "MSE computation bug"
  - id: FALSIFY-DIST-002
    rule: "Projection shape"
    prediction: "proj(student_hidden).shape == teacher_hidden.shape"
    if_fails: "Projection layer dimensions wrong"
  - id: FALSIFY-DIST-003
    rule: "Online/offline parity"
    prediction: "same batch → same teacher logits in both modes"
    if_fails: "Online mode has different teacher preprocessing"
  - id: FALSIFY-DIST-004
    rule: "Teacher frozen"
    prediction: "teacher weights bitwise identical before/after distillation"
    if_fails: "Optimizer accidentally includes teacher params"
  - id: FALSIFY-DIST-005
    rule: "Quant noise injection"
    prediction: "Q(x) != x for non-grid-aligned values"
    if_fails: "Quantization simulation is a no-op"
```

### Binding Requirements

Add to `contracts/entrenar/binding.yaml`:
```yaml
  - contract: distillation-strategies-v2.yaml
    equation: hidden_state_loss
    module_path: "entrenar::distill::hidden_state"
    function: hidden_state_distill_loss
    status: not_implemented

  - contract: distillation-strategies-v2.yaml
    equation: quant_aware_loss
    module_path: "entrenar::distill::quant_aware"
    function: quant_aware_distill_loss
    status: not_implemented

  - contract: distillation-strategies-v2.yaml
    equation: online_parity
    module_path: "entrenar::distill::online"
    function: online_distill
    status: not_implemented
```

### Verification Targets
- L2: 5 falsification tests
- L3: Kani harness for projection shape correctness
- L4: Lean proof for MSE non-negativity
